ciri/consensus/snapshot/local_snapshots/pruning_service.c

```diff
@@ -0,0 +1,333 @@
+                                               spent_addresses_provider_t *const sap,
+                                               flex_trit_t const *const entry_point_hash,
+                                               hash243_set_t *const transactions_to_prune);
+
+static retcode_t collect_unconfirmed_future_transactions_for_pruning_do_func(flex_trit_t *hash, iota_stor_pack_t *pack,
+                                                                             void *data, bool *should_branch,
+                                                                             bool *should_stop);
+
+static retcode_t collect_transactions_for_pruning_do_func(flex_trit_t *hash, iota_stor_pack_t *pack, void *data,
+                                                          bool *should_branch, bool *should_stop);
+
+static void *pruning_service_routine(void *arg);
+
+typedef struct collect_transactions_for_pruning_do_func_params_s {
+  uint64_t current_snapshot_index;
+  tangle_t const *const tangle;
+  hash243_set_t *const transactions_to_prune;
+} collect_transactions_for_pruning_do_func_params_t;
+
+static retcode_t collect_unconfirmed_future_transactions_for_pruning_do_func(flex_trit_t *hash, iota_stor_pack_t *pack,
+                                                                             void *data, bool *should_branch,
+                                                                             bool *should_stop) {
+  retcode_t ret = RC_OK;
+  *should_stop = false;
+  *should_branch = true;
+  collect_transactions_for_pruning_do_func_params_t *params = data;
+
+  if (pack->num_loaded == 0) {
+    *should_branch = false;
+    return RC_OK;
+  }
+
+  transaction_snapshot_index((iota_transaction_t *)pack->models[0]);
+
+  if (transaction_snapshot_index((iota_transaction_t *)pack->models[0]) == 0) {
+    if ((ret = hash243_set_add(params->transactions_to_prune, hash)) != RC_OK) {
+      return ret;
+    }
+    *should_branch = true;
+  }
+
+  return RC_OK;
+}
+
+static retcode_t collect_transactions_for_pruning_do_func(flex_trit_t *hash, iota_stor_pack_t *pack, void *data,
+                                                          bool *should_branch, bool *should_stop) {
+  retcode_t ret = RC_OK;
+  *should_stop = false;
+  *should_branch = false;
+  collect_transactions_for_pruning_do_func_params_t *params = data;
+
+  if (pack->num_loaded == 0) {
+    return RC_OK;
+  }
+
+  uint64_t current_snapshot_index = transaction_snapshot_index((iota_transaction_t *)pack->models[0]);
+
+  if (current_snapshot_index >= params->current_snapshot_index) {
+    *should_branch = true;
+    if ((ret = hash243_set_add(params->transactions_to_prune, hash)) != RC_OK) {
+      return ret;
+    }
+
+    ERR_BIND_RETURN(
+        tangle_traversal_dfs_to_future(params->tangle, collect_unconfirmed_future_transactions_for_pruning_do_func,
+                                       hash, NULL, &params),
```

Comment:
Shouldn't we give `params->transactions_to_prune` as `analyzed_hashes_param` instead of NULL here ?